sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java

```diff
@@ -707,111 +760,19 @@ public URL getContainerUrl() {
      * [!code-java[Sample_Code](../azure-storage-java/src/test/java/com/microsoft/azure/storage/Samples.java?name=list_blobs_hierarchy_helper "helper code for ContainerAsyncClient.listBlobsHierarchySegment")] \n
      * For more samples, please see the [Samples file](%https://github.com/Azure/azure-storage-java/blob/master/src/test/java/com/microsoft/azure/storage/Samples.java)
      */
-    private Mono<ContainersListBlobHierarchySegmentResponse> listBlobsHierarchySegment(String marker, String delimiter, ListBlobsOptions options) {
+    private Mono<ContainersListBlobHierarchySegmentResponse> listBlobsHierarchySegment(String marker, String delimiter,
+            ListBlobsOptions options, Duration timeout) {
```

Comment:
I don't think we generally accept timeout as a parameter on async apis. I'd also be surprised if we wanted this overload that returns ContainersListBlobHierarchySegmentResponse. I think we always want it to be PagedFlux, no?

Reply:
This is a private helper method for the paginated methods above. The `PagedFlux` implementation makes use of this to lazily fetch the next page.

Regarding the timeout, this is because of the synchronous wrappers for the paginated method; none of the public async methods accept a timeout. The synchronous methods for paginated listings must wrap the `PagedFlux` in a `PagedIterable`. However, if the user specifies a timeout for this operation, we cannot apply a timeout on the `PagedFlux` BEFORE the user decides if they want this by-item or by-page (`Flux::timeout(Duration)` returns `TimeoutFlux`, a hidden implementation of `Flux`, destroying our ability to use `PagedFlux` functionality). Since we cannot apply the timeout at the syncrhonous convenience layer, we need to find a new place. The chosen method here is diving down and applying the timeout to each individual `Mono` in the lazily-fetched pages. Considering `Flux::timeout(Duration)` applies the timeout between each individual emission, this is nearly identical to how a timeout would be applied to the `Flux` as a whole.